src/stlink-lib/usb.c

```diff
@@ -1159,6 +1159,40 @@ static stlink_backend_t _stlink_usb_backend = {
     _stlink_usb_read_trace
 };
 
+int stlink_serial(struct libusb_device_handle *handle, struct libusb_device_descriptor *desc, char *serial) {
+	unsigned char desc_serial[(STLINK_SERIAL_MAX_SIZE + 1) * 2];
+
+	/* get the LANGID from String Descriptor Zero */
+	int ret = libusb_get_string_descriptor(handle, 0, 0, desc_serial, sizeof(desc_serial));
+
+	if (ret < 4) return ret;
+
+	uint32_t langid = desc_serial[2] | (desc_serial[3] << 8);
+
+	/* get the serial */
+	ret = libusb_get_string_descriptor(handle, desc->iSerialNumber, langid, desc_serial,
+		sizeof(desc_serial));
+
+	if (ret < 0) return ret; // could not read serial
+
+	unsigned char len = desc_serial[0];
+
+	if (len == ((STLINK_SERIAL_MAX_SIZE + 1) * 2)) { /* len == 50 */
+		/* good ST-Link adapter */
+		ret = libusb_get_string_descriptor_ascii(
+			handle, desc->iSerialNumber, (unsigned char *)serial, STLINK_SERIAL_MAX_SIZE + 1);
```

Comment:
I would add escaping `serial` in this case as well. Although it's easier to move line 1188 to line before `return 0` (line 1193). The serial number is displayed by printf and this can lead to unpredictable behavior.

Reply:
not needed, read the manual
or dig in usb library
https://github.com/libusb/libusb/blob/eee6998395184d87bd8e9c07ce2637caed1207f4/libusb/descriptor.c#L1137

Reply:
Thanks for the info! I have not found in the doxygen that the returned string is \0 terminated